### DISCUSSION TITLE
fish 3.7.0

### DISCUSSION THREAD

**Discussion by zanchey** [fish-shell/fish-shell#9684] [General]
I've not used Discussions much, but let's give it a crack.

The Rust rewrite is well underway and it seems pretty clear that releasing a half-C++, half-Rust hybrid is an explicit non-goal.

If there are significant bugs in fish 3.6.1 that need to be fixed, we can produce a 3.6.2 release. As part of this, we can roll up minor                                                             
fixes, new and improved completions, and improvements to the documentation.

What I'm less sure on is whether it's worth planning a 3.6.2 (or call it a 3.7.0 branched from 3.6.1) in a few months with minor bugs, new completions, documentation improvements and any safe new features. We've traditionally avoided new features in point releases, just by necessity, but as we don't support anything except the latest branch I think it's probably effectively the same thing.

**Comment by faho** [comment]
As I see it, there are two separate reasons for making a 3.6.2/3.7.0:

Either we make it before the rust track is ready, because it's taking too long

or we make it because some users are having issues upgrading to the rust track and so we support them for a bit while they get their situation sorted.

For the first, I'm not sure how many features we'll even add - personally I've been holding back on doing any C++ work because I don't want to write things twice before they ever get used. So I guess we'll have to see how many important or easy bugs show up that we would want to backport the fixes for.

For the second, I've said before that we could do that but there would have to be interest. As I've seen it, most users should be able to build the rust version by just using rustup, so there is less of an argument than there was for sticking to C++11 (because upgrading C++ compilers is a PITA). So, for the most part, this would be targeting packagers.

In the past we've seen some packagers not upgrade even to patch releases (which is how e.g. 3.1.0 keeps turning up in bug reports), so I would like to see some interest in this before we waste our time.

(and if it's for packagers, it's probably okay to have it as a git tag only, I've seen that work for other projects - which would also mean no macos packages, which should ease releasing a bit)

**Reply by faho** [reply]
I have been tagging issues that would be okay for a bugfix 3.6.2: https://github.com/fish-shell/fish-shell/issues?q=is%3Aissue+milestone%3A%22fish+3.6.2%22

For now it's mostly completion fixes, together with a documentation improvement and a fix to the doc css - which isn't all that critical because it'll mostly be seen on mobile and I've updated the hosted version at fishshell.com.

A few of the [3.7.0 closed issues](https://github.com/fish-shell/fish-shell/issues?q=is%3Aissue+milestone%3A%22fish+3.7.0%22+is%3Aclosed) could also be added:

- #9699 - fix $PATH search for root
- #9659 - fix `exit` with negative returns
- #9497 - fossil vcs prompt
- #9452, completions
- #9443 - fish_vi_key_bindings in no-config mode
- #8285 - disable bracketed paste for `read`
- #4656 - an "external" cursor mode for vi-mode
- #3707 - don't ignore custom completions with leading "."

**Reply by faho** [reply]
#9814 is annoying - Python changed something in 3.12, and it prints warnings

**Reply by faho** [reply]
@zanchey Honestly I believe we have enough to warrant a 3.6.2, even just taking the 15 already fixed issues, which range from completion bugs (#9538, #9738, #9705) to a compatibility fix with python 3.12 (#9814).

I would also take at least #9699 and #9659 from the 3.7.0 issues.

**Reply by zanchey** [reply]
Yeah, let's do it.

**Reply by faho** [reply]
We should also add 7f76f75 and 1f1975689ea674f54a3de7a93e5dae018857e8ff. I neglected to open an issue for it, but it speeds up `git add ` completions if you have a lot of untracked files by a factor of ~5-8x.

**Comment by PatrickF1** [comment]
Sorry this is off topic--was it written anywhere how the fish team decided to rewrite in rust? Personally I'm excited because this will surely attract lots more contributors from the massive Rust CLI tooling out there. But wondering how y'all weighed the pros and cons. Thanks!

**Reply by faho** [reply]
Read #9512 - in short we were all, for various reasons, annoyed with C++ and ridiculousfish wasn't confident in switching to a multi-threading model in it even after months of work.

Rust is the main candidate there - it has the mind share, it has the model to help with data races, it has the ecosystem.

**Reply by PatrickF1** [reply]
Thanks for explaining! Good luck!

**Comment by zanchey** [comment]
I think we'll take most of those additional items, and call it 3.7.0.

**Comment by faho** [comment]
Alright, we're doing [3.7.0](https://github.com/fish-shell/fish-shell/tree/Integration_3.7.0), and here's the commits I've picked so far:

- cf67709 #9500, fossil
- af49b4d #8285, bracketed paste for read
- a16abf2 #9659, exit with negative value
- ca02e88 #9690, docs overflow for narrow screens
- 3a72d09 #9681, stack completions
- ba77858 #9664, md-to-clip completions
- c39780f #9538, __fish_complete_directories to remove --foo=
- 563b4d2 #9538, git
- bc04abe #9692, git
- e45bddc #9698, ignore unknown cursor shapes
- 9bd1dc1 #9647, zabbix completions
- c67d77f #9699, fix command completions as root
- d671710 #5319, docs on pipes
- 85504ca #9705, zfs completion error
- 0f1ef34 #9707, adb path
- b5bfff9 #9706, mix completions
- a6e16a1 #9713, docs
- 6ff971e #9721, docs typo
- 4ed53d4 #9717, apply foreground color for selection
- bbe2a2b #9722, completion descriptions
- 36e4b0f #9720, ar completion
- 8a0510a #9723, qjs completion
- fdd4bcf #9718, git switch remote branches
- bda9d57 #9732, ansible typo
- 9e22357 #9734, composer require
- db5c9ba #9738, git with '()' path
- f5e063a #9731, qjsc completions
- 5640390 #9728, docs for fish_X_key_bindings
- beca704 #9742/9741, git switch completions
- 30ae715 #9743, ip completions
- 480133b #9740, jq/gojq completions
- 20b500d #9669/9667, rpm-ostree
- 93dc848 #9750, remove kitty completions
- b76e6c5 #9753, don't autosuggest with bare $
- d855725 #9783, dnf lowercase
- 5f672ec #9787, don't read man page twice
- 67d1d80 #9794, look for tinfo where you look for curses
- aac3036 #9804, more systemctl subcommands
- 2016101 #9807, sync dracula theme
- 4e13b1b # Dracula licensing!
- 2eba684 #9814, fix python syntaxwarning
- ffb6168 #9813, age completions
- 6b1e6dd #9759, git update-index
- 272d123 #9825, doc typo
- b5fae430c0266468ca280d2001914190b375c672 #9824, german translations
- 6c6d28193839d8155fd2deafe630e962129997fe #9824
- a0a2475ccbd53b00f95676dac3a2ac5859fb17c2 #9824
- 71c320c #9833 / 9812 - sometimes completion pager skips selecting last entry
- 4c9fa51 #9817, don't use homebrew manpath
- cbf9a3b #9828, print css
- 38ac21b #8720, escape alias when printing
- bab8fb9 #9850, unzip i/o completion
- e3e7ab7 #9871, git show/diff stashes
- ac2810e #9878, rclone completions
- bfd97ad #9878, adding version parsing
- e31c0eb #9880, typo
- 72de1dc #9882, sphinx code block
- a6c36a0 #9888, let -c error out
- 6823f5e #9891, skip an access()
- 0cfdc9055132f2842007da95bf105a8d122141c3 #9850 fixup, fix for FreeBSD
- 6325b3662de122b571e33b7121cfc9d172d483bc #9899/9900 string repeat overflow
- 2bc605625e974e0adbc05603b77446a63d6fae7b #9904, gnu imp completions
- c56f9e198174cd740eed200deac677e322cce1ce #9915, doc fix
- 8d3885b9cbeaaac40b26e0510d23a5eac98683b7 #9905, blender completions
- 6ce2ffbbb0ec74f5ff57210215cb3629a453d370 #9903, krita completions
- 2110b364268d1f765e693e3a6ba6c5ae65afec10 #9914, gcc -O completions
- 9a9e133b184601ff838ef6bae825709adc227d19 #9919, gcc -lm completions
- ee75b4568723c5c260394a42cc41ab999ffd5b71 #9931, claw back some perf loss from 9699
- 21ddfabb8d2ff8180419230c7fac462604843ed5 #9939, zfs completions
- f9d21cc21d6559891fe44f0bc325c25d291a7253 #9922, horcrux completions
- 408ab860906fbf6e08f314bea982220fdee3428e #9932/6884, iwctl completions
- 69ef51f417cf30d396c33fce1c7cf0f05c5d5b4c #9955, turn on PWD reporting in iTerm
- c07136e8d3c07fb763d16f54502dcb934ac365c0 #9956, docs
- fd68aca6eabe3762baaa227640d0c9341e694527 #9954, ~ for alt-l
- 556bee6893a1cfd5af909aaaff2221999ce759fa #9960, more on iwctl
- 0f19d7118b10ee4354317888963c7291334e0173 #9961, more quotes in man parse
- 716001789b7789e71280868524fe860ae15daa0c #9970, docs
- 6473a9c763c867cd15ed022742e5081ad4cfaf88 #9965, completion description
- 81cd0359509621bd8fb11556bdfd65462ede54a9 #9972, fix apt completions being incomplete
- 5b1ff9459a4163fea275474f4cf92e273dc4f29e #9975, silence git in scales prompt
- e555f1b2355a059d115e7c505ae46bf15de281a6 #9983, docs
- b48fa1f1a0cf021ea714ae9e9b6b70aedc22825e #9982, docs css
- 482616f101cb3af25f36d8fcac7ca0bc9d78f7cb #9985, allow `foo | command time`
- b454b3bc40a2af3cf4a3034007c06a6d2a998f35 #9985, allow `foo | command and`
- 55c425a0ddcc6ebbdf26d25ef38de2043fc6d47a #9986, fish_key_reader key names
- f6123d235c1317132e0a36246605adba46a908f0 #9943 / 9704, scp path escaping
- 9d0d16686e07a8c7051071d33a83757b1f58da87 #9910, rpm-ostree completions
- 33ec25da8a88b12eaaf38c1850b6fe48ecc5ec7c #9998, java-home completions
- fff320b56ba42964033089809c77b38147058598 #9977 / 9976, make completions line continuations
- 0e81d25b369c202cf8e256382104552d4023921e # docs
- 5b44c26a197fd73a4aa465fa1d43bb4c67805fd0 #10005, use existing search string in pager
- 3bf80b23742717bd7f16804ae377bc72931eaaaa #10007, gcc completions libs
- d6e234c60caa7d17109c0637095be01920f10ef7 #10004, ibmcloud completions
- NOT APPLIED YET: #### 7534572d9953fa855ad75ad52502d6fec2014e24 #9994, node completion descriptions (didn't apply cleanly because of reformatting)
- b03327f5d2e0c7062f5af13d194e5c989df28c69 #10011, complete_command --foo= logic
- 3bcde90a88030ca29dfd3d97abf6d795a947dc31 #10023, loginctl completions
- 7c5777a82a3f78ce0b6256d163727b20d2bf0d0d #10029, pkill local variables
- 10d91b0249f57a5269ee5e1c0517da095fe3c820 #9617, pager colors lacking bold
- e6bef40c22758ea999540323bc362b8088d182c1 #10027, watchexec completions
- 269c9c3f0c0308c0abaaf716b8c5a5e22b92c274 #10034, crc/oc completions
- 0cdf801d0b979c1aa53310b905791a6514792eef #10039, pacman -Qp completions
- f8e38819a50367bcdd6a26fe7b18b21c1dffbe23 #10045, xdg-open with CLI apps
- ff433b0cb235689a93734e08a9dbed0204ce7ebe #7910, repaint-mode
- 496d65fb5d9660f2f0287a7c3951f96324f04de8 #9995, variable overrides for exec
- a8096724129e7246bddca0c7552e3ddac51b3b36 #10020, ls $opt scope
- c4ca1a68d361c2f0e636c2f5aba53fc0c35a52f1 #10044, clear-screen bind func
- 86803e44429b002d77342bdf8984dd790d5d7779 #10032, reduce stat for wildcards
- #10008

As well as these commits without an issue:

- ad54f073285e34e5eb63e3a02828aac5cae5be4a # Important doc fix for `set`.
- 098b7093da27775c2e0d788d66fc249021ce0658 # fish_config save with multiple values

**Reply by krobelus** [reply]
sounds good. I'd add some of the history pager changes if I get around to it

**Reply by faho** [reply]
Some more possible backports:

- #9635
- #9565
- #9776
- #9736
- #9841
- #9515
- #9875
- #9942
- #9913
- #9934

I would also *love* to have #9984 - unfortunately the fix we went with (passing an iterator that knows to skip underscores) isn't something I think we can make work in C++, so we'd have to go with the awkward solution I originally tried.

**Reply by krobelus** [reply]
yes and https://github.com/fish-shell/fish-shell/pull/9476

Yeah the C++ equivalent of Iterator::filter is https://en.cppreference.com/w/cpp/ranges/filter_view which is C++20

**Reply by rymrg** [reply]
Can you please backport this as well?

* [Improve fossil prompt execution time #9528](https://github.com/fish-shell/fish-shell/pull/9528)

**Reply by faho** [reply]
Labeled it now. We'll go through the closed issues/prs again and look for more to backport.

**Comment by zanchey** [comment]
"in a few months" was a bit optimistic but 3.7.0 is on the way soon
